integrationtests/tests/tests.rs

```diff
@@ -188,10 +191,133 @@ async fn lightning_gateway_pays_invoice() {
+    fed.run_consensus_epochs(2).await;
+
+    // Gateway deposits ecash to trigger preimage decryption by the federation
+    let (txid, contract_id) = gateway
+        .server
+        .buy_preimage_offer(invoice.payment_hash(), &amount, rng())
+        .await
+        .unwrap();
+    fed.run_consensus_epochs(5).await;
+
+    // Gateway receives decrypted preimage
+    let preimage = gateway
+        .server
+        .await_preimage_decryption(txid)
+        .await
+        .unwrap();
+
+    // Check that the preimage matches user pubkey & lightning invoice preimage
+    let pubkey = PublicKey::from_keypair(&secp(), &keypair);
+    assert_eq!(pubkey, preimage.0);
+    assert_eq!(&sha256(&pubkey.serialize()), invoice.payment_hash());
+
+    // User claims their ecash
+    user.client
+        .claim_incoming_contract(contract_id, keypair, rng())
+        .await
+        .unwrap();
+    fed.run_consensus_epochs(4).await;
+
+    // User fetches their coins
+    user.client.fetch_all_coins().await.unwrap();
+    fed.run_consensus_epochs(2).await;
+
+    // Ecash tokens have been transferred from gateway to user
+    assert_eq!(gateway.user.client.coins().amount(), sats(0));
+    assert_eq!(user.total_coins(), amount);
+}
+
+#[tokio::test(flavor = "multi_thread")]
+async fn receive_lightning_payment_invalid_preimage() {
+    let amount = sats(2000);
+    let (fed, user, _, gateway, _) = fixtures(2, 0, &[sats(10), sats(1000)]).await;
+    fed.mint_coins_for_user(&gateway.user, amount).await;
+    assert_eq!(user.total_coins(), sats(0));
+    assert_eq!(gateway.user.total_coins(), amount);
+
+    // Manually contstruct offer w/ invalid preimage. Can't use library code because it uses correct preimage ;)
+    let (fake_keypair, fake_pubkey) = secp().generate_schnorrsig_keypair(&mut rng());
+    let (real_keypair, real_pubkey) = secp().generate_schnorrsig_keypair(&mut rng());
+    let payment_hash = sha256(&real_pubkey.serialize());
+    let offer = IncomingContractOffer {
+        amount,
+        hash: payment_hash,
+        encrypted_preimage: EncryptedPreimage::new(
+            fake_pubkey.serialize(),
+            &user.config.ln.threshold_pub_key,
+        ),
+    };
+    let mut builder = TransactionBuilder::default();
+    builder.output(Output::LN(ContractOrOfferOutput::Offer(offer.clone())));
+    let tx = builder.build(&secp(), &mut rng());
+    fed.submit_transaction(tx);
+
+    // Run 2 epochs to get the offer to show up p
+    fed.run_consensus_epochs(2).await;
+
+    // Gateway escrows ecash to trigger preimage decryption by the federation
+    let (_, contract_id) = gateway
+        .server
+        .buy_preimage_offer(&payment_hash, &amount, rng())
+        .await
+        .unwrap();
+
+    // Now everyone has 0 ecash
+    assert_eq!(user.total_coins(), sats(0));
+    assert_eq!(gateway.user.total_coins(), sats(0));
+
+    fed.run_consensus_epochs(5).await;
+
+    // User gets error when they try to claim gateway's escrowed ecash
```

Comment:
I'm guessing the reason you test with the real / fake pubkeys are because you get different errors, maybe you can assert the error type using `assert_matches!`

Reply:
There actually wasn't much of a reason. I changed to just using one keypair and `sha256[0])` as the payment hash